Doc/howto/concurrency.rst

```diff
@@ -0,0 +1,1846 @@
+operating system will assign a certain set of limited resources to each
+process.  For example, each process has its own PID and handle to the
+executable.  You can run only so many processes before you run out of
+these resources.  Concurrency in a single process doesn't have this
+problem, and a distributed program can work around it.
+
+.. _multiprocessing-distributed:
+
+Using multiprocessing for distributed computing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Not only does the :mod:`multiprocessing` module support concurrency
+with multiple local processes, it can also support a distributed model
+using remote computers.  That said, consider first looking into tools
+that have been designed specifically for distributed computing,
+like :pypi:`dask`.
+
+Resilience to crashes
+^^^^^^^^^^^^^^^^^^^^^
+
+A process can crash if it does something it shouldn't, like try to
+access memory outside what the OS has provided it.  If your program
+is running in multiple processes (incl. distributed) then you can
+more easily recover from a crash in any one process.  Recovering
+from a crash when using free-threading, multiple interpreters, or
+coroutines isn't nearly so easy.
+
+High-level APIs
+---------------
+
+Also note that Python's stdlib provides various higher-level APIs
+that support these concurrency models in various contexts:
+
+.. list-table::
+   :header-rows: 1
+   :class: borderless vert-aligned
+   :align: left
+
+   * -
+     - :mod:`concurrent.futures`
+     - :mod:`socketserver`
+     - :mod:`http.server`
+   * - free-threading
+     - :class:`yes <concurrent.futures.ThreadPoolExecutor>`
+     - :class:`yes <socketserver.ThreadingMixIn>`
+     - :class:`yes <http.server.ThreadingHTTPServer>`
+   * - multiple interpreters
+     - (`pending <python-stdlib-interpreters_>`_)
+     -
+     -
+   * - coroutines
+     - ???
+     -
+     -
+   * - multi-processing
+     - | :class:`yes <concurrent.futures.ProcessPoolExecutor>`
+       | (:class:`similar <multiprocessing.pool.Pool>`)
+     - :class:`yes <socketserver.ForkingMixIn>`
+     -
+   * - distributed
+     - ???
+     -
+     -
+
+
+.. raw:: html
+
+   <br/>
+
+----
+
+.. _concurrency-design:
+
+Designing A Program For Concurrency
+===================================
+
+Whether you are starting a new project using concurrency or refactoring
+an existing one to use it, it's important to design for concurrency
+before taking one more step.  Doing so will save you a lot of
+headache later.
+
+1. decide if your program *might* benefit from concurrency
+2. `break down your *logical* program into distinct tasks <concurrency-identify-tasks_>`_
+3. `determine which tasks could run at the same time <concurrency-identify-tasks_>`_
+4. `identify the other concurrency-related characteristics of your program <concurrency-characteristics_>`_
+5. `decide which concurrency model fits best <concurrency-pick-a-model_>`_
+6. go for it!
```

Comment:
Surely to be done more explicit below, but maybe here we can hint that one of the major characteristics of your program regarding concurrency to select which model is best or even understand if concurrency actually applies is to identify if your program is CPU-bound or IO-bound.

Reply:
Great point.